File: sdks/elixir/lib/autonoma/template.ex
defmodule Autonoma.Template do
  @moduledoc """
  Template expression resolution for {{...}} expressions in entity specs.
  """

  @template_re ~r/\{\{(.+?)\}\}/

  @doc """
  Resolve all {{...}} expressions in a value.
  Handles strings, maps, lists recursively.
  """
  def resolve(value, ctx) when is_binary(value) do
    resolve_string(value, ctx)
  end

  def resolve(value, ctx) when is_list(value) do
    Enum.map(value, fn v -> resolve(v, ctx) end)
  end

  def resolve(value, ctx) when is_map(value) do
    Map.new(value, fn {k, v} -> {k, resolve(v, ctx)} end)
  end

  def resolve(value, _ctx), do: value

  defp resolve_string(str, ctx) do
    # If the entire string is a single expression, return the raw value (preserving type)
    case Regex.run(~r/^\{\{(.+?)\}\}$/, str) do
      [_, expr] ->
        evaluate_expression(String.trim(expr), ctx)

      nil ->
        # Interpolate expressions into the string
        Regex.replace(@template_re, str, fn _, expr ->
          expr |> String.trim() |> evaluate_expression(ctx) |> to_string()
        end)
    end
  end

  defp evaluate_expression("testRunId", ctx), do: ctx["testRunId"] || ctx[:testRunId]
  defp evaluate_expression("index", ctx), do: ctx["index"] || ctx[:index]

  defp evaluate_expression("index1", ctx) do
    index = ctx["index"] || ctx[:index]
    index + 1
  end

  defp evaluate_expression("now()", _ctx) do
    DateTime.utc_now() |> DateTime.to_iso8601()
  end

  defp evaluate_expression(expr, ctx) do
    cond do
      String.starts_with?(expr, "cycle(") ->
        items = parse_array_literal(expr, "cycle")
        index = ctx["index"] || ctx[:index]
        Enum.at(items, rem(index, length(items)))

      String.starts_with?(expr, "pick(") ->
        items = parse_array_literal(expr, "pick")
        Enum.random(items)

      String.starts_with?(expr, "random.int(") ->
        {min, max} = parse_range(expr, "random.int")
        min + :rand.uniform(max - min + 1) - 1

      String.starts_with?(expr, "random.float(") ->
        {min, max} = parse_range(expr, "random.float")
        :rand.uniform() * (max - min) + min

      String.starts_with?(expr, "daysAgo(") ->
        n = parse_single_int(expr, "daysAgo")
        DateTime.utc_now()
        |> DateTime.add(-n * 86400, :second)
        |> DateTime.to_iso8601()

      true ->
        raise "Template error: unknown expression '#{expr}'"
    end
  end

  defp parse_array_literal(expr, fn_name) do
    # Extract the array content: fn_name([...])
    inner =
      expr
      |> String.trim_leading("#{fn_name}([")
      |> String.trim_trailing("])")

    inner
    |> String.split(",")
    |> Enum.map(fn s ->
      s = String.trim(s)

      cond do
        String.starts_with?(s, "'") && String.ends_with?(s, "'") ->
          String.slice(s, 1..-2//1)

        String.starts_with?(s, "\"") && String.ends_with?(s, "\"") ->
          String.slice(s, 1..-2//1)

        true ->
          s
      end
    end)
  end

  defp parse_range(expr, fn_name) do
    inner =
      expr
      |> String.trim_leading("#{fn_name}(")
      |> String.trim_trailing(")")

    [a, b] = String.split(inner, ",") |> Enum.map(&String.trim/1)

    if fn_name == "random.int" do
      {String.to_integer(a), String.to_integer(b)}
    else
      {parse_number(a), parse_number(b)}
    end
  end

  defp parse_number(s) do
    case Float.parse(s) do
      {f, ""} -> f
      _ -> String.to_integer(s) * 1.0
    end
  end

  defp parse_single_int(expr, fn_name) do
    expr
    |> String.trim_leading("#{fn_name}(")
    |> String.trim_trailing(")")
    |> String.trim()
    |> String.to_integer()
  end
end

File: sdks/go/autonoma/template.go
package autonoma

import (
	"fmt"
	"math/rand"
	"regexp"
	"strconv"
	"strings"
	"time"
)

var templateRE = regexp.MustCompile(`\{\{(.+?)\}\}`)
var fullTemplateRE = regexp.MustCompile(`^\{\{(.+?)\}\}$`)

// TemplateContext holds the variables available to template expressions.
type TemplateContext struct {
	TestRunID string
	Index     int
}

// ResolveTemplate resolves all {{...}} expressions in a value.
// Handles strings, maps, and slices recursively.
func ResolveTemplate(value any, ctx TemplateContext) any {
	switch v := value.(type) {
	case string:
		return resolveString(v, ctx)
	case []any:
		result := make([]any, len(v))
		for i, item := range v {
			result[i] = ResolveTemplate(item, ctx)
		}
		return result
	case map[string]any:
		result := make(map[string]any, len(v))
		for k, val := range v {
			result[k] = ResolveTemplate(val, ctx)
		}
		return result
	default:
		return value
	}
}

func resolveString(str string, ctx TemplateContext) any {
	// If the entire string is a single expression, return the raw value (preserving type)
	if m := fullTemplateRE.FindStringSubmatch(str); m != nil {
		return evaluateExpression(m[1], ctx)
	}

	// Otherwise, interpolate expressions into the string
	return templateRE.ReplaceAllStringFunc(str, func(match string) string {
		inner := templateRE.FindStringSubmatch(match)
		if inner == nil {
			return match
		}
		val := evaluateExpression(inner[1], ctx)
		return fmt.Sprintf("%v", val)
	})
}

var (
	cycleRE     = regexp.MustCompile(`^cycle\(\[(.+)\]\)$`)
	pickRE      = regexp.MustCompile(`^pick\(\[(.+)\]\)$`)
	randIntRE   = regexp.MustCompile(`^random\.int\((\d+),\s*(\d+)\)$`)
	randFloatRE = regexp.MustCompile(`^random\.float\((\d+(?:\.\d+)?),\s*(\d+(?:\.\d+)?)\)$`)
	daysAgoRE   = regexp.MustCompile(`^daysAgo\((\d+)\)$`)
)

func evaluateExpression(expr string, ctx TemplateContext) any {
	expr = strings.TrimSpace(expr)

	switch expr {
	case "testRunId":
		return ctx.TestRunID
	case "index":
		return float64(ctx.Index)
	case "index1":
		return float64(ctx.Index + 1)
	case "now()":
		return time.Now().UTC().Format(time.RFC3339Nano)
	}

	// cycle([...])
	if m := cycleRE.FindStringSubmatch(expr); m != nil {
		items := parseArrayLiteral(m[1])
		return items[ctx.Index%len(items)]
	}

	// pick([...])
	if m := pickRE.FindStringSubmatch(expr); m != nil {
		items := parseArrayLiteral(m[1])
		return items[rand.Intn(len(items))]
	}

	// random.int(a,b)
	if m := randIntRE.FindStringSubmatch(expr); m != nil {
		min, _ := strconv.Atoi(m[1])
		max, _ := strconv.Atoi(m[2])
		return float64(rand.Intn(max-min+1) + min)
	}

	// random.float(a,b)
	if m := randFloatRE.FindStringSubmatch(expr); m != nil {
		min, _ := strconv.ParseFloat(m[1], 64)
		max, _ := strconv.ParseFloat(m[2], 64)
		return rand.Float64()*(max-min) + min
	}

	// daysAgo(n)
	if m := daysAgoRE.FindStringSubmatch(expr); m != nil {
		days, _ := strconv.Atoi(m[1])
		t := time.Now().UTC().Add(-time.Duration(days) * 24 * time.Hour)
		return t.Format(time.RFC3339Nano)
	}

	panic(fmt.Sprintf("Template error: unknown expression '%s'", expr))
}

func parseArrayLiteral(raw string) []string {
	parts := strings.Split(raw, ",")
	result := make([]string, len(parts))
	for i, s := range parts {
		s = strings.TrimSpace(s)
		// Strip surrounding quotes
		if (strings.HasPrefix(s, "'") && strings.HasSuffix(s, "'")) ||
			(strings.HasPrefix(s, "\"") && strings.HasSuffix(s, "\"")) {
			s = s[1 : len(s)-1]
		}
		result[i] = s
	}
	return result
}

